src/AddressId.sol

```diff
@@ -0,0 +1,131 @@
+        dripsHub.give(calcUserId(msg.sender), receiver, _calcAssetId(erc20), amt);
+    }
+
+    /// @notice Sets the msg.sender's drips configuration.
+    /// Transfers funds between the msg.sender's wallet and the drips hub contract
+    /// to fulfill the change of the drips balance.
+    /// @param erc20 The token to use
+    /// @param lastUpdate The timestamp of the last drips update of the user or the account.
+    /// If this is the first update, pass zero.
+    /// @param lastBalance The drips balance after the last drips update of the user or the account.
+    /// If this is the first update, pass zero.
+    /// @param currReceivers The list of the drips receivers set in the last drips update
+    /// of the user or the account.
+    /// If this is the first update, pass an empty array.
+    /// @param balanceDelta The drips balance change to be applied.
+    /// Positive to add funds to the drips balance, negative to remove them.
+    /// @param newReceivers The list of the drips receivers of the user or the account to be set.
+    /// Must be sorted by the receivers' addresses, deduplicated and without 0 amtPerSecs.
+    /// @return newBalance The new drips balance of the user or the account.
+    /// Pass it as `lastBalance` when updating that user or the account for the next time.
+    /// @return realBalanceDelta The actually applied drips balance change.
+    function setDrips(
+        IERC20 erc20,
+        uint64 lastUpdate,
+        uint128 lastBalance,
+        DripsReceiver[] memory currReceivers,
+        int128 balanceDelta,
+        DripsReceiver[] memory newReceivers
+    ) public returns (uint128 newBalance, int128 realBalanceDelta) {
+        if (balanceDelta > 0) _transferFromCaller(erc20, uint128(balanceDelta));
+        (newBalance, realBalanceDelta) = dripsHub.setDrips(
+            calcUserId(msg.sender),
+            _calcAssetId(erc20),
+            lastUpdate,
+            lastBalance,
+            currReceivers,
+            balanceDelta,
+            newReceivers
+        );
+        if (realBalanceDelta < 0) _transferTo(msg.sender, erc20, uint128(-realBalanceDelta));
+    }
+
+    /// @notice Sets msg.sender's splits configuration.
+    /// @param receivers The list of the user's splits receivers to be set.
+    /// Must be sorted by the splits receivers' addresses, deduplicated and without 0 weights.
+    /// Each splits receiver will be getting `weight / TOTAL_SPLITS_WEIGHT`
+    /// share of the funds collected by the user.
+    function setSplits(SplitsReceiver[] memory receivers) public {
+        dripsHub.setSplits(calcUserId(msg.sender), receivers);
+    }
+
+    function _calcAssetId(IERC20 erc20) internal pure returns (uint256) {
+        return uint160(address(erc20));
+    }
+
+    function _transferFromCaller(IERC20 erc20, uint128 amt) internal {
+        require(erc20.transferFrom(msg.sender, address(this), amt), "Transfer from caller failed");
+        erc20.approve(address(dripsHub), amt);
```

Comment:
I would expect only the transfer `_transferFromCaller` and not the approval for the next step here ;)

Reply:
I remember the discussion around the `registerAsset` and only do the approval once. On the other hand this makes the contract more complicated to use.

However, the `uint.max` approval wouldn't be a problem because the contract is not supposed to hold any ERC20 balance after a transaction.

Reply:
Yeah, the name is misleading. What should we call it, `_prepareForTransfer`? Maybe `_mediateTokens`?

This approach is the easiest to use and also it's very cheap. During a transaction we set here the ERC-20's `approvals` between addressId and dripsHub to `amt` and later dripsHub resets it to 0, so the cost of `approve` is amortized to the cost of a warm storage SLOAD via refunds. I agree that `registerAsset` is an overkill.

Maybe we should check the current approval and set it to `uint.max` if too low. After all even a noop approval emits events which cost gas and in 99.99% of calls approval will be already set making it almost always cheaper than the current approach.

Reply:
Updated:
```solidity
if (erc20.allowance(address(this), address(dripsHub)) < amt) {
    erc20.approve(address(dripsHub), type(uint256).max);
}
```